File: src/main/java/tri/le/music/security/AuthenticationTokenFilter.java
package tri.le.music.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import tri.le.music.service.AuthService;
import tri.le.music.util.TokenUtil;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

/**
 * Created by TriLe on 2016-04-03.
 */
@Component
public class AuthenticationTokenFilter extends OncePerRequestFilter {

    @Autowired
    AuthenticationManager authManager;

    @Autowired
    AuthService authService;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {

        String token = request.getHeader("token");

        if (TokenUtil.validate(token)) {
            // determine the user based on the (already validated) token
            UserDetails userDetails = authService.loadUserByToken(token);

            if (userDetails != null) {

                Authentication authentication = new UsernamePasswordAuthenticationToken(token, userDetails, userDetails.getAuthorities()); //this.authenticationProvider.authenticate(token);
                // set the authentication into the SecurityContext
                SecurityContextHolder.getContext().setAuthentication(authManager.authenticate(authentication));
            }
        }
        // next the filter chain
        filterChain.doFilter(request, response);
    }
}

File: src/main/java/tri/le/music/security/AuthenticationTokenFilter.kt
package tri.le.music.security

import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import tri.le.music.service.impl.AuthService
import tri.le.music.util.validate
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * Created by TriLe on 2016-04-22.
 */
@Component
class AuthenticationTokenFilter : OncePerRequestFilter() {

    @Autowired
    lateinit var authService: AuthService

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        val token = request.getHeader("token")

        if (validate(token)) {
            // determine the user based on the (already validated) token
            val userDetails = authService.loadUserByToken(token)

            if (userDetails != null) {

                val authentication = UsernamePasswordAuthenticationToken(token, userDetails, userDetails.authorities) //this.authenticationProvider.authenticate(token);
                // set the authentication into the SecurityContext
                SecurityContextHolder.getContext().authentication = authentication
            }
        }
        // next the filter chain
        filterChain.doFilter(request, response)
    }

    companion object {

        private val log = LogManager.getLogger(AuthenticationTokenFilter::class.java)
    }
}
